File: 110-balanced-binary-tree/balanced-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode* root) {
        return isBalancedDFS(root).first;
    }

    pair<bool, int> isBalancedDFS(TreeNode* root){
        if (! root){
            pair<bool,int> temp(true, 0);
            return temp;
        }

        pair<bool,int> left = isBalancedDFS(root->left);
        pair<bool,int> right = isBalancedDFS(root->right);
        bool balanced = abs(left.second - right.second) <= 1 && left.first && right.first ? true : false;
        pair<bool, int> result (balanced, 1 + max(left.second, right.second));
        return result;
    }
};

File: 110-balanced-binary-tree/balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:

    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        return self.isBalancedDFS(root)[0]

    def isBalancedDFS(self, root):
        if not root: return [True, 0]
        left = self.isBalancedDFS(root.left)
        right = self.isBalancedDFS(root.right)
        balanced = left[0] and right[0] and abs(left[1] - right[1]) <= 1
        return [balanced, 1+max(left[1], right[1])]
        